### ISSUE TITLE
[bug] recheck Argent77's mod Djinni companion works properly

### ISSUE THREAD

**Issue by lynxlynxlynx** [gemrb/gemrb#428]
We have an old report that [it](https://github.com/Argent77/DjinniCompanion/releases) doesn't display the 2nd quickweapon in the action bar properly ("no 2nd weapon displayed due to an unset value for qwslot2"). Recheck and retest if anything else is dodgy.

**Comment by caillean7** [comment]
That mod is quite extensive, but I take it you're mainly interested in the 'technical' stuff like those quickslots and maybe his spells and special abilities? Then I'd propose I could test all those and a couple of hours of gameplay for now. When I next make a serious playthrough, I'll be be sure to include this mod (sounds interesting anyway, I always thought Argent would do EE stuff only).

**Comment by lynxlynxlynx** [comment]
Yeah, this is just about these technicalities, so there's no real need for long untargeted playtesting. Summoning him and checking that he's set up properly is all it takes.

**Comment by caillean7** [comment]
I found 2 problems:

- the quickslots in fact don't work. On IE, if you summon the djinni, he has two weapon quickslots, in one he has a special enchanted scimitar, a basic unenchanted one in the other. On GemRB, both weapons do not show up in the quickslots. His avatar wields a weapon, just you can't switch them.

- after you summoned him for the first time (which works) and told him to return to his lamp, re-summoning him fails with the following console output:

```
[Control/WARNING]: Nested event handlers are not supported!
[Control/WARNING]: Nested event handlers are not supported!
[ResourceManager]: Found 'a7afqhlp.cre' in 'Override'.
[ResourceManager]: Found 'a7afqhlp.bcs' in 'Override'.
[ResourceManager/ERROR]: Couldn't find 'none.bam'.
[GameScript/ERROR]: ActionOverride failed for object: 
[GameScript/DEBUG]: Object: A7Afaaq

```
Spellcasting and his innate abilities seem to work fine.

**Comment by caillean7** [comment]
Some additional info:

The lamp has an effect (A7AFQHLP.eff) on it, that summons a helper creature (A7AQHLP.cre), whose [script](https://pastebin.com/NBrgegj8) (A7AFQHLP.bcs) takes care of summoning the Djinni (A7Afaaq.cre) himself.

**Comment by lynxlynxlynx** [comment]
That's a pretty odd looking sequence:
>         CreateCreature("A7Afaaq",[0.0],0)  // Afaaq
>         ActionOverride("A7Afaaq",MakeGlobal())
>         ActionOverride("A7Afaaq",DestroySelf())

**Comment by caillean7** [comment]
This block seems to be for the full version only - you never see this DisplayStringHead message if you're using the light version - as I previously did. I now installed the full version. And there's something else - if I use a 'djinni is in the bottle save' made by the IE, it works on GemRB to resummon him. It's only if I tell him to go into his lamp on GemRB that I can't resummon him.

**Comment by caillean7** [comment]
Telling to go to his lamp sets a local variable (Unsummoning) to 1, which seems to trigger the following block in his script:

```
IF
	!AreaCheck("A77006")
	HPGT(Myself,3)
	Global("A7AfaaqSummoned","GLOBAL",1)
	Global("A7AfaaqGaseous","GLOBAL",0)
	Global("unsummoning","LOCALS",1)
THEN
	RESPONSE #100
		SetGlobal("unsummoning","LOCALS",0)
		SetInterrupt(FALSE)
		ChangeEnemyAlly(Myself,NEUTRAL)
		ApplySpellRES("A7IMMUN","A7Afaaq")  // Immunity for 1s
		ApplySpellRES("A7UNSUM",Myself)  // Unsummon Afaaq
		SmallWait(5)
		SetSequence(SEQ_DIE)
		SmallWait(25)
		ChangeEnemyAlly(Myself,FAMILIAR)
		SetInterrupt(TRUE)
		SetGlobal("A7AfaaqSummoned","GLOBAL",0)
		SetGlobalTimer("A7AfaaqHealTimer","GLOBAL",ONE_TURN)
		SetGlobalTimer("A7AfaaqRestTimer","GLOBAL",EIGHT_HOURS)
		DestroySelf()
END
```

**Comment by lynxlynxlynx** [comment]
Hmm, nothing stands out. Try commenting out the timers — does it change anything? If not, just repeat and comment out more and more of this block. Only the first and last line are obviously necessary.

**Comment by caillean7** [comment]
:woman_shrugging: Even if comment everything out except for the first and last lines, it won't work.

**Comment by lynxlynxlynx** [comment]
Cool, that was one of the options. Sounds like the DestroySelf is too destructive ... but what is A7UNSUM for, just gfx?

**Comment by caillean7** [comment]
Yes, visual effect and destroy foot circle.

**Comment by lynxlynxlynx** [comment]
so if you leave out DestroySelf, does resummoning work?

**Comment by caillean7** [comment]
Yes! He gets invisible, but the foot circle is still shown, and the resummoning is all good.

**Comment by lynxlynxlynx** [comment]
Bleh. I hope they didn't have special hacks for EA of FAMILIAR. But I don't understand the first snippet I quoted either, is that really supposed to summon him (there's no other CreateCreature in the script you posted)? If so, then why the DestroySelf there?

I've pushed something, so we can get more info where the summoning fails ("ActionOverride failed for object: "). Please try again and check the log.

I've split the original stuff ...

**Comment by caillean7** [comment]
That's the only Create Creature in this script. I don't get it with DestroySelf almost immediately after CreateCreature. As for the log:

```
[ResourceManager/ERROR]: Couldn't find 'none.bam'.
[GameScript/ERROR]: ActionOverride failed for object and action: 
[GameScript/DEBUG]: Object: A7Afaaq

[GameScript/DEBUG]: Int0: 1, Int1: 0, Int2: 0
String0: localssummoning, String1: <NULL>
Point: [0.0]
1. Object: A7Afaaq
2. Object - NULL
3. Object - NULL
RefCount: 2	actionID: 30

```

**Comment by lynxlynxlynx** [comment]
Ok, so it's this line:
> ActionOverride("A7Afaaq",SetGlobal("summoning","LOCALS",1))

I don't think it would help, but you can try commenting out the preceding Deactivate call.

**Comment by caillean7** [comment]
Nope, commenting this line out doesn't help.

**Comment by lynxlynxlynx** [comment]
At this point I should just install it and check it with gdb. I know you have some familiarity with it, but since it's ActionOverride, breaking there will stop all the time, so it's a bit trickier to narrow down the correct call to start the actual debugging.

**Comment by caillean7** [comment]
To be honest, I used gdb just the couple of times when you told me what to do, so besides just saying 'backtrace' all the time I have no idea what to do. You probably get it quicker done yourself, instead of telling me what to do, yes...

**Comment by lynxlynxlynx** [comment]
While working on the inventory bug I noticed that the genie doesn't end up being a global actor. His script does a DestroySelf soon after creation and he's demoted back to a plain creature (likely the block I quoted above, didn't investigate). Things just go south from there, even though he appears to mostly work regardless.

**Comment by lynxlynxlynx** [comment]
Ok, so it is most likely a problem in the way we implemented MakeGlobal, which doesn't match IESDP. Argent77 doesn't remember the specifics of why he implemented the djinni this way, but it doesn't matter — it works. I'll open a fresh issue for MakeGlobal and once that's cleared up, this can be retested.

ref: https://github.com/Argent77/DjinniCompanion/issues/2
